eventuate-tram-sagas-reactive-orchestration-simple-dsl/src/main/java/io/eventuate/tram/sagas/reactive/simpledsl/ReactiveParticipantInvocationStep.java

```diff
@@ -0,0 +1,57 @@
+package io.eventuate.tram.sagas.reactive.simpledsl;
+
+import io.eventuate.tram.commands.common.ReplyMessageHeaders;
+import io.eventuate.tram.commands.consumer.CommandWithDestination;
+import io.eventuate.tram.messaging.common.Message;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+public class ReactiveParticipantInvocationStep<Data> implements ReactiveSagaStep<Data> {
```

Comment:
This class seems identical to https://github.com/eventuate-tram/eventuate-tram-sagas/blob/d01a4621e03b979069bde6cdc789cd1bc82bc2fd/eventuate-tram-sagas-orchestration-simple-dsl/src/main/java/io/eventuate/tram/sagas/simpledsl/ParticipantInvocationStep.java

except that, for example, returns Publisher<?> instead of Void.

Reply:
Chris, all fields contain reactive classes: https://github.com/eventuate-tram/eventuate-tram-sagas/blob/d01a4621e03b979069bde6cdc789cd1bc82bc2fd/eventuate-tram-sagas-reactive-orchestration-simple-dsl/src/main/java/io/eventuate/tram/sagas/reactive/simpledsl/ReactiveParticipantInvocationStep.java#L15-L18

and many methods contain reactive components:
https://github.com/eventuate-tram/eventuate-tram-sagas/blob/d01a4621e03b979069bde6cdc789cd1bc82bc2fd/eventuate-tram-sagas-reactive-orchestration-simple-dsl/src/main/java/io/eventuate/tram/sagas/reactive/simpledsl/ReactiveParticipantInvocationStep.java#L53
https://github.com/eventuate-tram/eventuate-tram-sagas/blob/d01a4621e03b979069bde6cdc789cd1bc82bc2fd/eventuate-tram-sagas-reactive-orchestration-simple-dsl/src/main/java/io/eventuate/tram/sagas/reactive/simpledsl/ReactiveParticipantInvocationStep.java#L42
https://github.com/eventuate-tram/eventuate-tram-sagas/blob/d01a4621e03b979069bde6cdc789cd1bc82bc2fd/eventuate-tram-sagas-reactive-orchestration-simple-dsl/src/main/java/io/eventuate/tram/sagas/reactive/simpledsl/ReactiveParticipantInvocationStep.java#L30

I don't think that any generalization is needed

Reply:
I'd suggest thinking about this some more.
For example, most of that class doesn't really care whether the types are reactive or not: only makeStepOutcome() 
Reactive actionReplyHandlers return Publisher<?>, non-reactive return Void - parameterizable.
You could explore moving most of the logic into a generic/parameterized superclass.

Also, consider:

```
public interface ReactiveParticipantInvocation<Data> {

  boolean isSuccessfulReply(Message message);

  boolean isInvocable(Data data);

  Publisher<CommandWithDestination> makeCommandToSend(Data data);
}
```

and

```
public interface ParticipantInvocation<Data> {

  boolean isSuccessfulReply(Message message);

  boolean isInvocable(Data data);

  CommandWithDestination makeCommandToSend(Data data);
}
```

They are really similar except for the return type of `makeCommandToSend()`


You could define 

```
public interface AbstractParticipantInvocation<Data, ReturnType> {

  boolean isSuccessfulReply(Message message);

  boolean isInvocable(Data data);

  ReturnType makeCommandToSend(Data data);
}

public interface ParticipantInvocation<Data> extends AbstractParticipantInvocation<Data, CommandWithDestination>

public interface ReactiveParticipantInvocation<Data> extends AbstractParticipantInvocation<Data, Publisher<CommandWithDestination>>

```

I'd imagine that the abstract superclass could be written in terms of AbstractParticipantInvocation

Reply:
Ok, but I think it is optimization, for now I think is important just to make it work